Swatomation

ahaessly opened the pull request:
merging testcode into ui code
- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: If you changed a URL that is used elsewhere (e.g. in an email), comment about where it is used and ensure the dependent code is updated.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] **Submitter**: If you're adding new libraries, sign us up to security updates for them
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Product Owner** sign off
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

ahaessly committed:
add test ids

asingh7115 committed:
DSDEQA-80: all the data properties

asingh7115 committed:
DSDEQA-80: adding more data-test-id's

asingh7115 committed:
DSDEQA-80: adding more data-test-id's

asingh7115 committed:
DSDEQA-80: moar data test id's

breilly2 committed:
Render extra Button props on the <a> tag. Add data-test-id to create workspace button.

breilly2 committed:
Render extra Button props on the <a> tag. Add data-test-id to create workspace button.

breilly2 committed:
PR feedback

breilly2 committed:
PR feedback

breilly2 committed:
Fixed logical merge conflict.

breilly2 committed:
Added data-test-id for account-dropdown and sign-out

asingh7115 committed:
adding sign-in button test-id

asingh7115 committed:
adding open-create-workspace-modal-button test-id

asingh7115 committed:
adding billing-project-select test-id

asingh7115 committed:
rebase create.cljs

breilly2 committed:
Fixed a data-test-id for method config importing

asingh7115 committed:
rebase

breilly2 committed:
Removed accidentally duplicated block of code

asingh7115 committed:
adding test ids for deleting a workspace

asingh7115 committed:
Launch analysis test ids

breilly2 committed:
Added support for applying data props to table rows.
Added various test IDs.

breilly2 committed:
Added test IDs.

breilly2 committed:
Merge remote-tracking branch 'origin/swatomation' into swatomation

Conflicts:
	src/cljs/main/broadfcui/page/workspace/summary/tab.cljs

asingh7115 committed:
Moving scala automated tests to ui repo

asingh7115 committed:
Changes to scala file structure

asingh7115 committed:
application.conf

asingh7115 committed:
restructure some stuff

ansingh7115 committed:
Merge pull request #822 from broadinstitute/swatomation-r

restructure some stuff

asingh7115 committed:
more restructure - src inside automation

asingh7115 committed:
update build.sbt to reflect new location of automated tests

jacmrob committed:
run through docker against fiab

asingh7115 committed:
merge develop

ansingh7115 committed:
Merge pull request #830 from broadinstitute/swatomation-new

Swatomation merge develop

jacmrob committed:
templatize conf and insert into tests when running

jacmrob committed:
couple things

ahaessly committed:
merge in new auth token and some qa users

asingh7115 committed:
test for cloning workspace

asingh7115 committed:
Merge branch 'swatomation' of https://github.com/broadinstitute/firecloud-ui into swatomation

ahaessly committed:
removed info

breilly2 committed:
Improved and added many more authorization domain tests.
Other minor cleanup.

asingh7115 committed:
adding import data test, changes to method config tests

ahaessly committed:
update tests to use new users from config

asingh7115 committed:
fixing withWebDriver for local and headless chrome

ahaessly committed:
changes to try to get tests to run headless remotely

ahaessly committed:
change dev to qa for thurloe url

jacmrob committed:
udpate ctmpls

ahaessly committed:
put / in front of chrome path

jacmrob committed:
add pem ctmpl and fix template render stuff

jacmrob committed:
fix failure screenshot rendering

jacmrob committed:
don't use ivy cache

jacmrob committed:
fix more SSL stuff

andy7i committed:
test id tags for method configs

added automation fixtures for abort submission

adding billing and method test ids

test code for workspace methods and billing projects

ahaessly committed:
Get tests running with qa users (#834)

* merge in new auth token and some qa users
* update tests to use new users from config
* get current tests to pass with new qa users
* change dev to qa for thurloe url
* Save screenshot on test failure.
* udpate ctmpls
* put / in front of chrome path
* add pem ctmpl and fix template render stuff
* documentation for automation
* fix failure screenshot rendering
* Added clean-up to registration test and migrated it to the new test users.
Added passing of FireCloud-Id to Thurloe calls.
Fixed a bug with xpath of checking for an element with text.
* don't use ivy cache
* add dsde-toolbox pull to runtests script
* Added Thurloe service
* changing local orch api
* use different auth domain
* get default auth domain from vault
* add host name to test runner
* fixed loading of particpants.txt for DataTabSpec
* fail whole script if tests fail
* Improved reliability of test for creating a billing project.
Don't hide the test failure if there is a problem capturing a screenshot.
Avoid stalled headless chrome nodes (https://github.com/SeleniumHQ/docker-selenium/issues/87).
* Re-throw error when logging.
Be a little more quiet about clean-up failures.
* Avoid instability with Google sign-in when the popup window automatically closes.

ahaessly committed:
Merge branch 'develop' into swatomation

ahaessly committed:
fix testing to work with new method import screens

breilly2 committed:
Capture page source as well as a screenshot on failure.

breilly2 committed:
Added pre-test clean-up to RegistrationSpec.

ahaessly committed:
Merge branch 'swatomation' of https://github.com/broadinstitute/firecloud-ui into swatomation

breilly2 committed:
Enabled full stacktraces from scalatest instead of manually intercepting and logging.

jacmrob committed:
write out hub and chrome1 logs

ahaessly committed:
change gestures to ui

ahaessly committed:
refactor publish tests

jacmrob committed:
fix

jacmrob committed:
Merge branch 'swatomation' of github.com:broadinstitute/firecloud-ui into swatomation

jacmrob committed:
write out logs from syslog driver for selenium chrome nodes

breilly2 committed:
More catching and suppressing errors during test tear-down to avoid hiding the real failure.

jacmrob committed:
change port (515 already in use by stackdriver on most hosts)

jacmrob committed:
Merge branch 'swatomation' of github.com:broadinstitute/firecloud-ui into swatomation

jacmrob committed:
fix port mapping

jacmrob committed:
pull selenium containers before running (to catch any updates)

ahaessly committed:
wait until submission completes before checking status

ahaessly committed:
change gestures to ui

ahaessly committed:
use new method with required inputs

new error check for missing input

jacmrob committed:
generalize confs to run tests locally, update documentation

ahaessly committed:
fixed publish tests
added modals

breilly2 committed:
Updated AuthDomainSpec to create its own groups to use as auth domains.
Various clean-up, documentation, etc.

breilly2 committed:
Merge remote-tracking branch 'origin/swatomation' into swatomation

rtitle commented:
Is this ready for review? I'm interested and would like to go through it (focusing mostly on the Scala stuff), but only if its authors say it's ready.

breilly2 commented:
@rtitle Thank you for offering to review. I'm hesitant to have you look at it just yet because there is some remaining code clean-up on my list that will improve consistency and readability. I expect to get through enough of my list by start-of-business tomorrow (Wednesday) to make it worth reviewing. I'll keep you updated.

breilly2 committed:
Removed unreliable `await toggle`.

breilly2 committed:
Changed `Option[String] = None` parameter to `String = ""` to reduce verbosity.

breilly2 committed:
Added more use of withWorkspace.
Removed need for type parameter when using WorkspaceFixtures.

andy7i committed:
all of the tests

breilly2 committed:
Changed `Option[String] = None` parameter to `String = ""` to reduce verbosity.

jacmrob committed:
remove authDomain from conf

jacmrob committed:
fix documentation and render local script to get rid of authDomain var

breilly2 committed:
Reorganized auth domain tests to clarify the 4 combinations of in/not in auth domain and shared/not shared.

breilly2 committed:
Merge remote-tracking branch 'origin/swatomation' into swatomation

andy7i committed:
applying review comments

andy7i committed:
Merge branch 'swatomation' into aj_convert_regression

andy7i committed:
Merge pull request #849 from broadinstitute/aj_convert_regression

converting existing regression tests into scala

jacmrob committed:
test pr maker

jacmrob committed:
test PR automation take 2

jacmrob committed:
test PR builder again

jacmrob committed:
one more time for the ppl in the back

andy7i committed:
fallout from merge

andy7i committed:
Merge branch 'swatomation' of https://github.com/broadinstitute/firecloud-ui into swatomation

breilly2 committed:
Moved many files into more logical places.

andy7i committed:
cleaning up / pr review comments

andy7i committed:
separated email field into its own div

breilly2 commented on src/cljs/main/broadfcui/header.cljs:
```diff
@@ -42,7 +42,7 @@
                             :padding "0.5rem" :border style/standard-line}
                     :data-test-id "account-dropdown"}
-              (-> auth2 (.-currentUser) (.get) (.getBasicProfile) (.getEmail))
+              [:div {:style {:display "inline-block"} :data-test-id "account-dropdown-email"} (-> auth2 (.-currentUser) (.get) (.getBasicProfile) (.getEmail))]
```
Any reason not to use a `span` instead of a `div`? The `span` should be inherently inline so you wouldn't need the `inline-block` style.

andy7i committed:
`span`s are inline blocks by default

rtitle commented on README.md:
````diff
@@ -68,3 +68,48 @@ To compile and build the `broadinstitute/firecloud-ui` docker image, run
+sbt clean
+```
+
+TODO - running specific tests/suites
````
You can run single suites like:
```
sbt "testOnly *GoogleSpec"
```
Or single tests within a suite like:
```
# matches test via substring
sbt "testOnly *GoogleSpec -- -z \"have a search field\""
```
See: http://www.scala-sbt.org/0.13/docs/Testing.html#Test+Framework+Arguments

rtitle commented on automation/src/test/scala/GoogleSpec.scala:
```diff
@@ -0,0 +1,26 @@
+
+  it should "have a search field" in withWebDriver { implicit driver =>
+    go to new GooglePage()
+    assert(find("q").isDefined)
```
I think you mentioned this in email, but I generally prefer ScalaTest matchers over `assert(boolean)` because it gives nicer error messages (and usually reads better). If you mix in the `ShouldMatchers` trait this could be written as:
```
find("q") shouldBe 'defined
``` 

See: http://www.scalatest.org/user_guide/using_matchers

Not really a big deal and I won't comment about other occurences of this.

breilly2 replied:
Looks like `ShouldMatchers` is deprecated in favor of `Matchers` which we are using in some of the other tests.

Thanks for the tip on using `defined` with matchers! I wasn't sure how to do that and I'm pretty sure there are some places where I can put that to good use.

rtitle replied:
Ah right, it's just `Matchers` now.

rtitle commented on automation/src/test/scala/GoogleSpec.scala:
```diff
@@ -0,0 +1,26 @@
+    go to new GooglePage()
+    val button = find(xpath("//input[@value='Google Search'][@type='submit']"))
+    assert(button.isDefined)
+    //    Thread sleep 10000
```
Remove commented code

rtitle commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/api/APIException.scala:
```diff
@@ -0,0 +1,4 @@
+package org.broadinstitute.dsde.firecloud.api
+
+
+class APIException (message: String = null, cause: Throwable = null) extends Exception(message, cause)
```
This could be a `case class`

rtitle commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/api/FireCloudClient.scala:
```diff
@@ -0,0 +1,113 @@
+class FireCloudClient {
+  implicit val system = ActorSystem()
+  implicit val materializer = ActorMaterializer()
+  implicit val ec: ExecutionContextExecutor = ExecutionContext.global
```
I've been telling people to avoid using the global execution context and use the `ActorSystem.dispatcher` instead. It probably doesn't matter much for tests. But I would still probably prefer:
```
implicit val ec: ExecutionContext = system.dispatcher
```

rtitle commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/api/FireCloudClient.scala:
```diff
@@ -0,0 +1,113 @@
+import scala.concurrent.duration._
+import scala.concurrent.{Await, ExecutionContext, ExecutionContextExecutor, Future}
+
+class FireCloudClient {
```
Looking below, `FireCloudClient` is extended by other traits. So it might be better to make `FireCloudClient` a trait itself. (I didn't even realize that a trait could extend a class: https://stackoverflow.com/questions/12854941/why-can-a-scala-trait-extend-a-class).

breilly2 replied:
Actually, I'm not sure why the things below this (`Orchestration`, `Rawls`, `Thurloe`) are traits with companion objects instead of just straight-up objects. @rtitle can you think of any reason not to make `FireCloudClient` a trait and the rest objects?

rtitle replied:
sounds reasonable to me

rtitle commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/api/FireCloudClient.scala:
```diff
@@ -0,0 +1,113 @@
+  val mapper = new ObjectMapper()
+  mapper.registerModule(DefaultScalaModule)
+
+  // TODO: Is this okay here or should it be in a separate object and explicitly imported?
```
I'm fine with it here: it will automatically be available to subclasses. Could mark it `protected` so _only_ subclasses can access it (in general I like narrowing down implicit scope as much as possible).

rtitle commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/api/FireCloudClient.scala:
```diff
@@ -0,0 +1,113 @@
+    }
+  }
+
+  def parseResponse(response: HttpResponse): String = {
```
What's the difference between `handleResponse` and `parseResponse`? They look identical to me.

Also I feel like this could be achieved more simply by acting on futures instead of chaining multiple `Await.result` calls. My attempt (annotated):
```
 private def sendRequestAndWaitForResponse(httpRequest: HttpRequest): String = {
    val future: Future[String] = for {
      response <- Http().singleRequest(httpRequest)  // makes the request and returns a Future[HttpResponse]
      responseAsString <- response.entity.toStrict(5 seconds).map(_.data.decodeString("UTF-8"))  // Loads the response into memory and decodes it as a String, returning a Future[String]
    } yield {
      // Throw exception if the response is a failure, else return it
      if (response.status.isFailure()) {
        throw new APIException(s"Request failed with status ${response.status.intValue}: $responseAsString")
      } else {
        responseAsString
      }
    }
    // Either returns a string or throws an exception
    Await.result(future, 5 minutes)
  }
```

breilly2 replied:
That is specifically on my list of things to clean up. If I remember correctly, the differences are subtle and have to do with needing to add support for something (I don't remember what) and feeling like it would require minor changes throughout the tests that I wasn't prepared to do just yet. Thanks for the implementation suggestion!

breilly2 replied:
We lost some git history here, but I think code in `handleResponse` used to be inline in `requestWithJsonContent` or something like that.

@rtitle, the only issue with your proposed `sendRequestAndWaitForResponse` is that callers sometimes want to inspect the response rather than deal with one of a `String` response or an `APIException`. I like how it acts on futures, though, so I'm trying to work it into the API.

breilly2 committed:
Changed CleanUp to run clean-up functions in the reverse order than they were registered.

rtitle commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/api/FireCloudClient.scala:
```diff
@@ -0,0 +1,113 @@
+    val formData = Multipart.FormData {
+      Source {
+        List (
+          Multipart.FormData.BodyPart.apply(name, HttpEntity(ByteString(content)))
```
Instead of `Source(List(Multipart.FormData.BodyPart(...)))` just `Source.single(Multipart.FormData.BodyPart(...))`. 

Can also omit the `.apply`

rtitle commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/api/Orchestration.scala:
```diff
@@ -0,0 +1,185 @@
+
+trait Orchestration extends FireCloudClient with LazyLogging {
+
+  def responseAsList(response: String): List[Map[String, Object]] = {
```
I wonder if this type of thing could be baked into the `parseResponse` in the superclass.  So you could do something like
```
val response: List[String] = parseResponse[List[Map[String, Object]]](getRequest(Config.FireCloud.orchApiUrl + "api/profile/billing"))
```

rtitle replied:
Something like:
```
  def parseResponse2[A: ClassTag](response: HttpResponse): A = {
    val f = response.entity.toStrict(5 seconds).map(entity => mapper.readValue(entity.data.toArray, classTag[A].runtimeClass.asInstanceOf[Class[A]]))
    Await.result(f, 100 millis)
  }
```
The `.asInstanceOf` is unfortunate but that's what you get with Jackson (I'm trying to avoid saying use a more idiomatic scala JSON parser :)).

breilly2 replied:
I played around with this a bit and didn't really like anything I came up with. The `asInstanceOf` bothered me, as did the verbosity of specifying the type at the call site for something that would likely be very common. I'd be willing to give it another try though.

rtitle commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/api/Orchestration.scala:
```diff
@@ -0,0 +1,185 @@
+    mapper.readValue(response, classOf[List[Map[String, Object]]])
+  }
+
+  private def retry[T](remainingBackOffIntervals: Seq[FiniteDuration])(op: => () => Boolean): Boolean = {
```
I'm not sure if this is duplicated elsewhere but this looks like some code that could be shared instead of a `private def`.

rtitle replied:
Also `op: => () => Boolean` seems wrong. There are too many layers of indirection. This is a "by-name parameter which returns a function which takes no arguments and returns a Boolean". But all you need is a "by-name parameter which returns a Boolean". So:
```
private def retry[T](remainingBackOffIntervals: Seq[FiniteDuration])(op: => Boolean): Boolean = {
  op match {
    // snip
  }
}
```
This should give you the behavior you need. More info: https://tpolecat.github.io/2014/06/26/call-by-name.html

rtitle replied:
Also the method is parameterized so it should be `op: => T`

breilly2 replied:
Good feedback. Yes, this does currently also live in `Util.scala`, though I've had it in mind to pull it into a `Retry.scala` instead. Both are copied from/inspired by Rawls code. In fact, the version in `Util` has the changes that you suggested. It's possible that one was added before the scala class and the other was added after. :)

rtitle commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/api/Orchestration.scala:
```diff
@@ -0,0 +1,185 @@
+    op() match {
+      case true => true
+      case false if remainingBackOffIntervals.isEmpty => false
+      case false if remainingBackOffIntervals.nonEmpty =>
```
Log messages might be nice here. E.g.:
```
log.error(s"Operation failed after $n retries")
log.warning(s"Operation failed, retrying attempt $n/$m")
etc
```

Rawls has a similar retry function: https://github.com/broadinstitute/rawls/blob/develop/util/src/main/scala/org/broadinstitute/dsde/rawls/util/Retry.scala#L47

rtitle commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/api/Orchestration.scala:
```diff
@@ -0,0 +1,185 @@
+
+  object billing {
+
+    object BillingProjectRole extends Enumeration {
```
Just noting the Scala example for `Enumeration`: http://www.scala-lang.org/api/2.12.0/scala/Enumeration.html

It includes some type aliasing which essentially lets you refer to enum values without typing `.Value` all the time. But this should work fine as well.

breilly2 replied:
Is that 2.12 only? I think we're on 2.11. I read up on enumeration patterns before I chose this one. Should be try to be on 2.12 if it would be easy to move?

rtitle replied:
It's just the 2.12 docs, scala.Enumeration has existed forever

breilly2 replied:
Okay. I can add that. I think I may have backed away from the type aliasing because it wasn't intuitive to me how to find usages of the enum in IntelliJ. I see now that you have to find usages of the type rather than the object which I guess makes sense.

rtitle commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/api/Orchestration.scala:
```diff
@@ -0,0 +1,185 @@
+    def createBillingProject(projectName: String, billingAccount: String)(implicit token: AuthToken): Unit = {
+      postRequest(Config.FireCloud.orchApiUrl + "api/billing", Map("projectName" -> projectName, "billingAccount" -> billingAccount))
+
+      retry(Seq.fill(30)(10.seconds)) { () =>
```
Also if you take my retry advice above you would write this as 
```
retry(Seq.fill(30)(10.seconds)) { 
  // code to retry
}
```

rtitle replied:
Also if retrying N times with a fixed delay is a common pattern it might be good to define retry as taking a number of attempts (`Int`) and a delay (`FiniteDuration`) which might be simpler to understand.

rtitle commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/api/Orchestration.scala:
```diff
@@ -0,0 +1,185 @@
+          e.exists(_ == ("creationStatus", "Ready")) && e.exists(_ == ("projectName", projectName)))
+      } match {
+        case false => throw new Exception("Billing project creation did not complete")
+        case true => println("Finished creating billing project")
```
Would be nice to use a logger such as logback-over-slf4j instead of `println`

rtitle commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/api/Orchestration.scala:
```diff
@@ -0,0 +1,185 @@
+    }
+
+    def addUserToGroup(groupName: String, email: String, role: GroupRole.Value)(implicit token: AuthToken): Unit = {
+      logger.info(s"Adding user to group: $groupName $email")
```
Oh here's a logger :)

rtitle commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/api/Orchestration.scala:
```diff
@@ -0,0 +1,185 @@
+              (implicit token: AuthToken): Unit = {
+      logger.info(s"Creating workspace: $namespace/$name with auth domain: $authDomain")
+
+      val authDomainMap = authDomain match {
```
Maybe `authDomain` should be a `Option[String]`

breilly2 replied:
I had that originally but didn't like how it looked when calling. Either the argument would be omitted or it would always be `Option["blah"]` which seemed redundant to me. The code is currently split between the 2 options but I definitely want to consistently use one or the other. Is there a strong argument for using an `Option`?

rtitle replied:
It's a scala best practice to use `Option` for optional values. IMO there's nothing wrong with the below code. I would say the `Some("domain")` parameter is actually useful because it informs the caller (via the type!) that the parameter is optional. Otherwise if it were a `String` the caller would be left guessing whether to put `""`, or `null`, etc..

```
def create(namespace: String, name: String, authDomain: Option[String] = None)
          (implicit token: AuthToken): Unit = {
  val authDomainMap = authDomain match {
    case None => Map()
    case Some(a) => Map("authorizationDomain" -> Map("membersGroupName" -> a))
  }
  ...
}

// Later:
create("namespace", "name", Some("domain"))
```

There's a lot of literature about why to use Option. Here's a random one I just found via Google: http://danielwestheide.com/blog/2012/12/19/the-neophytes-guide-to-scala-part-5-the-option-type.html

breilly2 replied:
Gotcha. I still feel like using `Option` and a default parameter value is redundant -- if you don't have one, you don't have to pass in anything -- but I guess using `Some(...)` isn't too bad.

rtitle commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/api/Orchestration.scala:
```diff
@@ -0,0 +1,185 @@
+
+  object library {
+    def setLibraryAttributes(ns: String, name: String, attributes: Map[String, Any])(implicit token: AuthToken): String = {
+      putRequest(Config.FireCloud.orchApiUrl + "api/library/" + ns + "/" + name + "/metadata", attributes)
```
Seems like a good candidate for string interpolation

rtitle commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/api/Rawls.scala:
```diff
@@ -0,0 +1,35 @@
+      getRequest(url + s"api/admin/user/$subjectId").status match {
+        case StatusCodes.OK => Option(true)
+        case StatusCodes.NotFound => Option(false)
+        case _ => None
```
Ah the old "troolean".

The only usage I see of this is doing a `.getOrElse(false)` -- maybe just make the method return a `Boolean`?

breilly2 replied:
The current use of this doesn't care if the result is undetermined (not `OK` or `NotFound`). However, I didn't want to assume that to be true of any future uses. I'm still working on getting the hang of when it's better to use `Option`.

rtitle commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/config/AuthTokens.scala:
```diff
@@ -0,0 +1,58 @@
+import com.google.api.client.googleapis.javanet.GoogleNetHttpTransport
+import com.google.api.client.json.jackson2.JacksonFactory
+
+import scala.collection.JavaConversions._
```
I think the best practice is to import `scala.collection.JavaConverters._` instead of `JavaConversions._`. The difference is that `JavaConverters` enriches Java/Scala collection classes with `.asJava` and `.asScala` methods, while `JavaConversions` tries to do it for you behind the scenes. It tends to be better to be explicit about when you want a conversion to avoid surprises/unexpected errors.

rtitle commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/config/AuthTokens.scala:
```diff
@@ -0,0 +1,58 @@
+      .setServiceAccountPrivateKeyFromPemFile(pemfile)
+      .setServiceAccountScopes(authScopes)
+      .setServiceAccountUser(userEmail)
+      .build();
```
semicolon

rtitle commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/config/AuthTokens.scala:
```diff
@@ -0,0 +1,58 @@
+      .setJsonFactory(jsonFactory)
+      .setServiceAccountId(email)
+      .setServiceAccountPrivateKeyFromPemFile(pemfile)
+      .setServiceAccountScopes(authScopes)
```
Probably would do `authScopes.asJava` if using `JavaConverters`.

rtitle commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/fixture/TestData.scala:
```diff
@@ -0,0 +1,62 @@
+
+  private def createSamples(samplePrefix: String, participantId: String, numberOfSamples: Int): String = {
+    "entity:sample_id\tparticipant_id\n" +
+      new Array[String](numberOfSamples).zipWithIndex.map { el => samplePrefix + el._2 + "\t" + participantId + "\n" }.mkString.stripSuffix("\n")
```
Doesn't seem like you're using the index so `zipWithIndex` is unneeded?

rtitle replied:
Another way: `(0 until numberOfSamples).map(i => ...)`

rtitle commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/AuthenticatedPage.scala:
```diff
@@ -0,0 +1,54 @@
+    def clickAccountDropdown(): Unit = {
+      click on accountDropdown
+    }
+we
```
rogue `we`

rtitle commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/AuthenticatedPage.scala:
```diff
@@ -0,0 +1,54 @@
+  }
+
+  /*
+   * This must be private so that subclasses can provide their own object
```
Not sure this is true? This `ui`'s class name would be `org.broadinstitute.dsde.firecloud.page. AuthenticatedPage.ui` while any subclasses' would be `org.broadinstitute.dsde.firecloud.page.TheSubClass.ui` so it shouldn't conflict.

breilly2 replied:
IntelliJ's static analysis doesn't complain, but compiling subclasses that have an `object ui ...` results in an error:
```
Error:(133, 10) overriding object ui in class AuthenticatedPage;
 object ui cannot override final member
  object ui extends UI
```

rtitle replied:
oh, ok

rtitle commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/PageUtil.scala:
```diff
@@ -0,0 +1,31 @@
+  def open(implicit webDriver: WebDriver): P = {
+    go to this
+    awaitLoaded()
+    this
```
Could `this` be removed (and just return the result of `awaitLoaded()`)?

rtitle commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/WorkspaceDataPage.scala:
```diff
@@ -0,0 +1,113 @@
+    }
+
+    def uploadData(filePath: String) = {
+      val strThatWorked = "/Users/ansingh/Desktop/ui-11/firecloud-ui/automation/src/test/scala/org/broadinstitute/dsde/firecloud/data/participants.txt"
```
Remove

rtitle commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/WorkspacePage.scala:
```diff
@@ -0,0 +1,13 @@
+A Workspace Page is any page within the workspace (i.e. the Summary tab, Data tab)
+ */
+class WorkspacePage(implicit webDriver: WebDriver) extends AuthenticatedPage {
+  
```
Any shared code that could live here? I think all the urls start with `/workspaces/$namespace/$name`, right?

rtitle commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/WorkspaceSummaryPage.scala:
```diff
@@ -0,0 +1,224 @@
+  /**
+    * Dictionary of access level labels displayed in the web UI.
+    */
+  object AccessLevel extends Enumeration {
```
I feel like I already saw this enum somewhere else -- move to a shared location?

breilly2 replied:
This `AccessLevel` encodes the display values in the UI. The other one that you saw was `Orchestration.WorkspaceAccessLevel` which encodes the values that the API works with. Although the string values are the same in this case, they are conceptually different things. In fact, I wish that the display values were not YELLING.

rtitle commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/WebBrowserSpec.scala:
```diff
@@ -0,0 +1,129 @@
+        val fileName = s"failure_screenshots/${date}_$suiteName.png"
+        val htmlSourceFileName = s"failure_screenshots/${date}_$suiteName.html"
+        try {
+          val tmpFile = new Augmenter().augment(driver).asInstanceOf[TakesScreenshot].getScreenshotAs(OutputType.FILE)
```
Crazy, scala code that takes a screenshot :)

rtitle commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/billing/BillingSpec.scala:
```diff
@@ -0,0 +1,124 @@
+          submissionDetailsPage.waitUntilSubmissionCompletes()
+          assert(submissionDetailsPage.verifyWorkflowSucceeded())
+        }
+
```
Great tests!

rtitle commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/metadata/MethodConfigTabSpec.scala:
```diff
@@ -0,0 +1,249 @@
+  }
+
+
+  // negative tests
```
who needs negative tests :)

rtitle commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/security/AuthDomainSpec.scala:
```diff
@@ -0,0 +1,199 @@
+ * This test SHOULD be able to run with ParallelTestExecution. However, Rawls
+ * currently has database deadlock problems when creating and deleting managed
+ * groups because they involve 3 operations across 2 tables. My initial
+ * attempt to retry the deadlock failed because Rawls also creates the groups
```
Interesting that this causes Rawls deadlocks - I wonder if this is a reproduction of https://broadinstitute.atlassian.net/browse/GAWB-2256..

breilly2 replied:
I think `GoogleGroupSyncMonitorSupervisor` is about reading the groups as they're defined in Rawls and updating Google to match. Therefore, it should only be reading from the group tables.

However, I think the queue of synchronization actions is actually in the database at some point so maybe that's the cause of GAWB-2256. Either way, I think we're quickly approaching the need to audit how we handle database operations and transactions which will result in a non-trivial amount of tech debt cleanup.

rtitle replied:
+1 on auditing our database access logic, some of the code is a little scary

rtitle commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/util/Util.scala:
```diff
@@ -0,0 +1,36 @@
+    UUID.randomUUID().toString
+  }
+
+  def retry[T](remainingBackOffIntervals: Seq[FiniteDuration])(op: => Option[T]): Option[T] = {
```
Hey here's a common `retry`! See my comments above.

rtitle commented on build.sbt:
```diff
@@ -0,0 +1,10 @@
+
+version := "1.0"
+
+scalaVersion := "2.11.8"
```
Ah... we should probably be using scala 2.12.2 for any new projects. Was there a specific reason to using 2.11?

breilly2 replied:
Just copied from other FireCloud projects. I guess this would be a good time to make a move toward 2.12.

rtitle replied:
I guess the Scala upgrade will come in a future iteration?

breilly2 replied:
Yes, unfortunately updating scala, scalatest, and project/dependency clean-up will have to come later. I definitely want to do that soon, though.

rtitle commented on project/Dependencies.scala:
```diff
@@ -0,0 +1,53 @@
+    "ch.qos.logback" % "logback-classic" % "1.1.3",
+    "com.google.apis" % "google-api-services-oauth2" % "v1-rev112-1.20.0",
+    "com.google.api-client" % "google-api-client" % "1.22.0" exclude("com.google.guava", "guava-jdk5"),
+//    "com.getsentry.raven" % "raven-logback" % "7.8.6",
```
Remove commented dependencies, this is really hard to read

rtitle commented on project/Dependencies.scala:
```diff
@@ -0,0 +1,53 @@
+//    "org.ocpsoft.prettytime" % "prettytime" % "4.0.1.Final",
+//    "org.everit.json"      %  "org.everit.json.schema" % "1.4.1",
+    "org.specs2"          %%  "specs2-core"   % "3.7"  % "test",
+    "org.scalatest"       %%  "scalatest"     % "2.2.6"   % "test",
```
scalatest 3 would be nice too

rtitle commented on project/Settings.scala:
```diff
@@ -0,0 +1,55 @@
+import Dependencies._
+//import Merging._
```
Please remove commented code in this file too

rtitle commented on src/cljs/main/broadfcui/auth.cljs:
```diff
@@ -62,7 +62,7 @@
      ;; than removing it from the page.
      [:div {:style {:display (when (:hidden? props) "none") :marginTop "2rem"}}
-      [comps/Button {:text "Sign In" :onClick #(this :-handle-sign-in-click)}]
+      [comps/Button {:text "Sign In" :onClick #(this :-handle-sign-in-click) :data-test-id "sign-in-button"}]
```
A more UI-literate person should probably review the cljs changes

rtitle commented:
I made it! The tests look great. I made a bunch of comments, mostly in util/test code.

andy7i committed:
Merge branch 'swatomation' of https://github.com/broadinstitute/firecloud-ui into swatomation

andy7i committed:
adding and using apis for creating/copying methods from method repo

andy7i committed:
temporarily reverting api use

andy7i committed:
Merge pull request #860 from broadinstitute/aj_methodconfig_api

methodconfig api

breilly2 committed:
Improved reliability of creating a billing project due to the create modal not going away right away.

breilly2 committed:
Merge remote-tracking branch 'origin/swatomation' into swatomation

andy7i committed:
using api

breilly2 committed:
Improved reliability of creating a billing project due to the create modal not going away right away.

andy7i committed:
fixed api

andy7i committed:
Merge pull request #862 from broadinstitute/aj_methodconfig_api

methodconfig api

breilly2 committed:
Minor PR feedback.

breilly2 committed:
Minor PR feedback.

jacmrob committed:
test PR automation

breilly2 committed:
More recovery from StaleElementReferenceException.

andy7i committed:
update readme on how to run single suite / testcase

zarsky-broad commented on Dockerfile-tests:
```diff
@@ -0,0 +1,12 @@
+FROM hseeberger/scala-sbt
```
Consider using an Alpine image. It'll take less space and less RAM, and it'll likely run faster.

zarsky-broad replied:
(This goes for all Docker images, if there's an Alpine version or you can make one yourself, life will likely be better…)

breilly2 replied:
@zarsky-broad Is there a trusted source for these images? I didn't see an obvious one to trust (though I guess I'm equally skeptical of hseeberger and I don't know why we chose it here).

zarsky-broad replied:
You might consider something like https://hub.docker.com/r/gafiatulin/alpine-sbt/, or you could roll your own. Since the Dockerfiles are public, there's no need for an "official" one as long as you're ok with how it's built.

zarsky-broad commented on src/cljs/main/broadfcui/common.cljs:
```diff
@@ -356,7 +356,8 @@
                              :onMouseOut #(swap! state assoc :hovering? false)
-                             :onClick (:dismiss props)}
+                             :onClick (:dismiss props)
+                             :data-test-id (:data-test-id props)}
```
At this point, I'm going to recommend that you `:keys` out `props`.

breilly2 replied:
@zarsky-broad can you check this to make sure I did it right? Thanks.

zarsky-broad commented on src/cljs/main/broadfcui/common/components.cljs:
```diff
@@ -558,12 +565,13 @@
+        [:div (merge {:style {:borderBottom style/standard-line
                        :padding "20px 48px 18px"
                        :fontSize "137%" :fontWeight 400 :lineHeight 1}}
+                     (when (some? data-test-id) {:data-test-id data-test-id}))
```
Is this `when some?` necessary? If a prop is nil, react will just not render it.

breilly2 replied:
Good call. Looks like the outer `merge` isn't needed either.

zarsky-broad reviewed (changes requested):
I think I'm going to ask that all `:data-test-id` values be wrapped in a `(when config/debug?)`.
This only has to be done on `defc`-ed components themselves and on raw html elements, but it should prevent the prop getting rendered to production.

breilly2 committed:
PR feedback and clean-up.

breilly2 committed:
Prevent chrome nodes from timing out after 30 seconds.
Changed some billing tests to create billing projects using the web service API.

breilly2 committed:
Attempt to improve detection of when the data library page has finished loading.

breilly2 committed:
Changed JavaConversions to JavaConverters.

breilly2 committed:
Minor clean-up and improvement from PR feedback.

breilly2 committed:
":keys out props"

breilly2 committed:
Simplification

breilly2 committed:
Changed `param: String = ""` to `param: Option[String] = None`.

breilly2 committed:
Changed `param: String = ""` to `param: Option[String] = None`.

breilly2 committed:
Merge branch 'develop' into swatomation.
Also fixed some merge fallout.

Conflicts:
	src/cljs/main/broadfcui/common/entity_table.cljs
	src/cljs/main/broadfcui/common/management_utils.cljs
	src/cljs/main/broadfcui/common/table/body.cljs
	src/cljs/main/broadfcui/page/billing/billing_management.cljs
	src/cljs/main/broadfcui/page/workspace/create.cljs
	src/cljs/main/broadfcui/page/workspace/data/tab.cljs
	src/cljs/main/broadfcui/page/workspace/details.cljs
	src/cljs/main/broadfcui/page/workspace/method_configs/launch_analysis.cljs
	src/cljs/main/broadfcui/page/workspace/method_configs/method_config_editor.cljs
	src/cljs/main/broadfcui/page/workspace/monitor/submission_details.cljs
	src/cljs/main/broadfcui/page/workspace/summary/tab.cljs
	src/cljs/main/broadfcui/page/workspace/summary/workspace_cloner.cljs
	src/cljs/main/broadfcui/page/workspaces_list.cljs

breilly2 committed:
Fixed post-merge problems.

breilly2 committed:
Fixed use of Typeahead component on method config page.

breilly2 committed:
Wait for sign-in popup to load.

breilly2 commented on src/cljs/main/broadfcui/page/workspace/monitor/common.cljs:
```diff
@@ -41,25 +41,29 @@
   (icons/icon {:style {:color (:success-state style/colors) :marginRight 4
-                       :width table-style/table-icon-size :height table-style/table-icon-size}}
+                       :width table-style/table-icon-size :height table-style/table-icon-size}
+               :data-test-id "status-icon" :data-test-value "success"}
```
I can not put `(config/when-debug ...)` around these values because, for some reason unknown to me, the config is consistently not yet loaded when these are rendered. Putnam suggested adding a watch to the config atom but I don't see how that's going to work in this context. I'm going to need some additional help fixing this. It would be great to fix this before merging but I'm not going to hold up the merge for this one case.

breilly2 committed:
Added `config/when-debug` around data-test-ids.
Minor reliability improvements for a couple of tests.

rtitle commented:
It looks like my comments have been addressed. 👍 !

zarsky-broad committed:
use defn's for status icons so debug is evaluated at render

breilly2 committed:
no comment

breilly2 committed:
Merge branch 'develop' into swatomation

Conflicts:
	src/cljs/main/broadfcui/page/billing/billing_management.cljs
	src/cljs/main/broadfcui/page/workspaces_list.cljs
